File: src/78-subsets.ts
function subsets(nums: number[]): number[][] {
  let result: number[][] = [];
  let queue: number[] = [];

  function backtrack (start = 0) {
    result.push(Array.from(queue));

    for (let i = start; i < nums.length; i++) {
      queue.push(nums[i]);
      backtrack(i + 1);
      queue.pop();
    }
  }

  backtrack();

  return result;
};

File: src/78-subsets.java
class Solution {
  private List<List<Integer>> result = new ArrayList<>();
  private List<Integer> queue = new ArrayList<>();

  public List<List<Integer>> subsets(int[] nums) {
    backtrack(0, nums);
    return result;
  }

  public void backtrack(int start, int[] nums) {
    result.add(new ArrayList<>(queue));

    for (int i = start; i < nums.length; i++) {
      queue.add(nums[i]);
      backtrack(i + 1, nums);
      queue.remove(queue.size() - 1);
    }
  }
}
